samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -0,0 +1,347 @@
+        "image_uri": image_uri,
+        "s3_bucket_name": bucket_name,
+    }
+
+    try:
+        rendered_template_string = Template(jinja_template_json_string).render(data)
+        return json.loads(rendered_template_string)
+    except Exception as render_error:
+        LOG.exception("Failed to render jinja template: %s", render_error)
+        return None
+
+
+def __query_tagging_api(tag_filters: dict) -> dict:
+    """
+    Queries the Tagging API to retrieve the ARNs of every resource with the given tags.
+
+    NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#get-resources
+
+    Parameters
+    ----------
+    tag_filters : dict
+        The tag filters to restrict output to only those resources with the given tags.
+
+        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#options
+
+    Returns
+    -------
+    dict
+        A response object containing ResourceTagMappingList, which contains a list of resource ARNs and tags associated with each.
+
+        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#output
+
+
+    """
+    try:
+        return boto3.client("resourcegroupstaggingapi").get_resources(tag_filters=tag_filters)
+    except Exception as query_error:
+        LOG.error("Failed to call get_resources from resource group tagging api: %s", query_error)
+        return None
+
+
+def __generate_statement_list(tag_filters: dict) -> List[dict]:
+    """
+    Generates a list of IAM statements with default action permissions for resources with the given tags.
+
+    Parameters
+    ----------
+    tag_filters : dict
+        The tag filters to restrict output to only those resources with the given tags.
+
+        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#options
+
+    Returns
+    -------
+    List[dict]
+        The list of JSON objects representing IAM statements corresponding to the resources specified by the given tags.
+    """
+
+    tagging_api_response = __query_tagging_api(tag_filters)
+
+    if tagging_api_response is None:
+        LOG.error("Failed to receive get_resources response, cannot generate any IAM statements.")
+        return []
+
+    iam_statements = []
+
+    for resource in tagging_api_response["ResourceTagMappingList"]:
+        new_statement = {
+            "Effect": "Allow",
+            "Action": [],
+            "Resource": "",
+        }
+        arn = resource["ResourceARN"]
+
+        new_statement["Action"].extend(
+            # Default permissions are commented out
+            ["# " + action for action in __get_permissions(arn)]
+        )
+
+        # https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html
+        if "# execute-api:Invoke" in new_statement["Action"]:
```

Comment:
Getting the `Action` and `Resource` for an ARN are coupled, but here in an indirect way that relies on the output of `__get_permissions`, instead of the ARN.

Could we instead have a function that returns the full IAM statement for some ARN? Should help make this loop cleaner as well.

Reply:
Refactored so that an ARN maps to the `AWS::Service::Resource` as well as the IAM action list. Now the whole IAM statement is generated in a single function, and it is more clear which Resource Type the ARN regex maps to. APIGW resources can also be dealt with directly in the statement generating function.